codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.handler.codec.compression;
+
+import com.aayushatharva.brotli4j.decoder.DecoderJNI;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+
+/**
+ * Uncompresses a {@link ByteBuf} encoded with the brotli format.
+ *
+ * See <a href="https://github.com/google/brotli">brotli</a>.
+ */
+public final class BrotliDecoder extends ByteToMessageDecoder {
+
+    private enum State {
+        DONE, NEEDS_MORE_INPUT, ERROR
+    }
+
+    static {
+        try {
+            Brotli.ensureAvailability();
+        } catch (Throwable throwable) {
+            throw new ExceptionInInitializerError(throwable);
+        }
+    }
+
+    private final int inputBufferSize;
+    private DecoderJNI.Wrapper decoder;
+    private boolean destroyed;
+
+    public BrotliDecoder() {
+        this(8 * 1024);
+    }
+
+    public BrotliDecoder(int inputBufferSize) {
+        this.inputBufferSize = inputBufferSize;
+    }
+
+    private ByteBuf pull(ByteBufAllocator alloc) {
+        ByteBuffer nativeBuffer = decoder.pull();
+        ByteBuf copy = alloc.buffer(nativeBuffer.remaining());
+        copy.writeBytes(nativeBuffer);
```

Comment:
Please, provide an inline comment for future developers with explanation why the content of returned `nativeBuffer` have to be copied, and can not be wrapped as-is or as `nativeBuffer.duplicate()`.

Reply:
done